lib/render/FlowRenderer.cpp

```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
+            const vector<flow::Particle> &stream = adv->GetStreamAt(s);
+            sv.clear();
+            int sn = stream.size();
+            if (_cache_isSteady)
+                sn = std::min(sn, (int)maxSamples);
+            
+            for (int i = 0; i < sn + 1; i++) {
+                // "IsSpecial" means don't render this sample.
+                if (i == sn || stream[i].IsSpecial()) {
+                    int svn = sv.size();
+                    
+                    if (svn < 2) {
+                        sv.clear();
+                        continue;
+                    }
+                    
+                    vec3 prep(-normalize(sv[1].p-sv[0].p) + sv[0].p);
+                    vec3 post( normalize(sv[svn-1].p-sv[svn-2].p) + sv[svn-1].p);
+                    
+                    size_t vn = vertices.size();
+                    vertices.resize(vn + svn + 2);
+                    vertices[vn] = {prep, sv[0].v};
+                    vertices[vertices.size()-1] = {post, sv[svn-1].v};
+                    
+                    memcpy(vertices.data() + vn + 1, sv.data(), sizeof(Vertex) * svn);
+                    
+                    sizes.push_back(svn+2);
+                    sv.clear();
+                } else {
+                    const flow::Particle &p = stream[i];
+                    
+                    if (_cache_isSteady) {
+                        sv.push_back({p.location, p.value});
+                    } else {
+                        if(p.time > _timestamps.at(_cache_currentTS))
+                            continue;
+                        if(p.time >= startingTime)
+                            sv.push_back({p.location, p.value});
+                    }
+                }
+            }
+            
+            _renderStatus = FlowStatus::UPTODATE;
+        }
+
+        assert(glIsVertexArray(_VAO) == GL_TRUE);
+        assert(glIsBuffer(_VBO) == GL_TRUE);
+
+        glBindVertexArray(_VAO);
+        glBindBuffer(GL_ARRAY_BUFFER, _VBO);
+        glBufferData(GL_ARRAY_BUFFER, sizeof(Vertex) * vertices.size(), vertices.data(), GL_STREAM_DRAW);
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+        
+        _streamSizes = sizes;
+    }
+    
+    
+    FlowParams::RenderType renderType = (FlowParams::RenderType)rp->GetValueLong(FlowParams::RenderTypeTag, FlowParams::RenderTypeStream);
+    FlowParams::GlpyhType glyphType = (FlowParams::GlpyhType)rp->GetValueLong(FlowParams::RenderGlyphTypeTag, FlowParams::GlpyhTypeSphere);
+    bool geom3d = rp->GetValueLong(FlowParams::RenderGeom3DTag, false);
+    bool show_dir = rp->GetValueLong(FlowParams::RenderShowStreamDirTag, false);
+    
+    float radiusBase = rp->GetValueDouble(FlowParams::RenderRadiusBaseTag, -1);
+    if (radiusBase == -1) {
+        vector<double> mind, maxd;
+        _dataMgr->GetVariableExtents(rp->GetCurrentTimestep(), rp->GetColorMapVariableName(), rp->GetRefinementLevel(), rp->GetCompressionLevel(), mind, maxd);
```

Comment:
I hope this doesn't come across as petty.  I've had difficulty in other parts of the code base because I need to resize my window, or reformat long lines to make sense of them.

A per-line character limit was documented in our Coding Convention, but we haven't been paying much attention to these rules.  I've recently violated some myself.  I think we should 1) try to enforce these, and 2) change them when needed.  

_7. Limit all lines to 80 characters whenever possible. Avoid deeply nested clauses and limit the number of lines in function/method to approximately one "page". Use two space indentation or tabs. Be consistent throughout the file.  
https://github.com/NCAR/VAPOR/wiki/Vapor-Coding-Convention_

An 80 character limit may be excessive.  This would mess with your header files that I think are nicely formatted, ie FlowRenderer.h.  A 100 character limit may strike a balance between making lines like this more readable, without requiring drastic changes to your code base.

Reply:
I think this is something that affects our entire code base. Let's talk about it during our meeting.

Reply:
I certainly think 80 character is a limit that's too small, thus I support increase it to a bigger value (100, 120, etc.).

At the same time I share the concern that a line that grows to an excessive length dramatically reduces its readability, and makes debugging difficult. 

As a side note, adopting tools such as clang-format and/or clang-tidy is a good solution for this type of hassles.

Reply:
Done. Just thought I'd mention that you can configure your text editors to wrap at a certain character limit, optionally keeping words whole and text aligned with the current indentation. This way everyone can have their own preferred line width.